keps/sig-instrumentation/1661-event-api-migration/README.md

```diff
@@ -0,0 +1,422 @@
+
+##### Short Examples:
+
+After first occurrence, Event looks like:
+```
+{
+  regarding: A,
+  action: B,
+  reportingController: C,
+  ...,
+}
+```
+After second occurrence, Event looks like:
+```
+{
+  regarding: A,
+  action: B,
+  reportingController: C,
+  ...,
+  series: {count: 2, state: "Ongoing"},
+}
+```
+After half an hour of crashlooping, Event looks like:
+```
+{
+  regarding: A,
+  action: B,
+  reportingController: C,
+  ...,
+  series: {count: 4242, state: "Ongoing"},
+}
+```
+Minute after crashloop stopped, Event looks like:
+```
+{
+  regarding: A,
+  action: B,
+  reportingController: C,
+  ...,
+  series: {count: 424242, state: "Finished"},
+}
+```
+
+#### Client Side Changes
+
+All clients will need to eventually migrate to use new Events, but no other actions are required from them. Deduplication logic change will be completely transparent after the move to the new API.
+
+#### Restarts
+
+Event recorder library will list all Events emitted by corresponding components and reconstruct internal activeSeries map from it.
+
+### Defence in Depth
+
+Because Events proved problematic we want to add multiple levels of protection in the client library to reduce chances that Events will be overloading API servers in the future. We propose to do two things.
+
+#### Aggressive Backoff
+
+We need to make sure that kubernetes client used by EventRecorder uses properly configured and backoff pretty aggressively. Events should not destabilize the cluster, so if EventRecorder receives 429 response it should exponentially back off for non-negligible amount of time, to let API server recover.
+
+### Other Related Changes
+
+To allow easier querying we need to make following fields selectable for Events:
+- event.reportingComponent
+- event.reportingInstance
+- event.action
+- event.reason
+- event.regarding...
+- event.related...
+- event.type
+
+Kubectl will need to be updated to use new Events if present.
+
+### Test Plan
+
+Correctness:
+
+- Update all unit tests that are using Event lib to use the new API and make sure they all pass.
+- Run tests with both healthy and crash-looping clusters that keep generating Events to ensure they are produced in an expected way.
+
+Scalability and Performance:
+
+- Record the memory usage in both healthy and crash-looping clusters of various size (e.g., 50, 500, 5k nodes).
+- Ensure memory increase can be handled by server and there's no big performance reduction.
+
+### Graduation Criteria
+
+The new Event API is in Beta right now. The plan is to graduate to GA in 1.19.
+
+#### Beta -> GA Graduation
+
+- All components are migrated to use the new Event API.
```

Comment:
I don't think we want to bundle those two together.
We already migrated scheduler and I think that made us confident that the API makes sense.

I think we should:
- target GA of the API itself in 1.19
- but not necessary require that all components will also be updated (we may try, but that shouldn't be a requirement)

@liggitt - do you have any specific concerns here from API perspective?

Reply:
I've updated this section also the summary. So to clarify, the main purpose currently for this KEP is to have API graduation to GA, and migrating components are also wanted but not the priority focus in this KEP right now, correct?

Reply:
Agree on decoupling API graduation from propagating use to all components.